### PR TITLE
[visualization] Updates node texture when mode/role changes

### PR DESCRIPTION
**Pull request by ctan-g**
This PR changes node shape update logic.
- Update node texture when role changes.
- Add support for `mode` status messages and update node texture when mode changes.
- Change router shape to hexagon to better distinguish node roles.

### CODE REVIEW

**Comment by qilinkang** [comment]
Looks nice and easy to distinguish node roles. Thanks for the quick turnaround.
Please update the codelabs in openthread.io after the PR is merged.
https://codelabs.developers.google.com/codelabs/openthread-network-simulator/#4

**Comment by simonlingoogle** [comment]
Use a different shape for leader?

**Comment by ctan-g** [comment]
> Use a different shape for leader?

Leader is in the same tier as other routers and colored differently (similar to OpenThread.io graphs). I feel it's okay to use same shape.

**Comment by simonlingoogle** [comment]
> > Use a different shape for leader?
> 
> Leader is in the same tier as other routers and colored differently (similar to OpenThread.io graphs). I feel it's okay to use same shape.

Fine to me.

File: types/types.go [hunk 1]
```diff
@@ -57,6 +60,19 @@ func DefaultNodeMode() NodeMode {
 	}
 }
 
+func ParseNodeModeString(s string) NodeMode {
```
Comment (simonlingoogle) [inline types/types.go:63]:
```suggestion
func ParseNodeMode(s string) NodeMode {
```
We can already tell `s` is a string from the argument type.

File: types/types.go [hunk 2]
```diff
@@ -57,6 +60,19 @@ func DefaultNodeMode() NodeMode {
...
+	RxOnWhenIdle := strings.Contains(s, "r")
+	SecureDataRequests := strings.Contains(s, "s")
+	FullThreadDevice := strings.Contains(s, "d")
+	FullNetworkData := strings.Contains(s, "n")
```
Comment (simonlingoogle) [inline types/types.go:67]: This code is less efficient than ideal. What about code like this
```go
for _, c := range s {
  switch c {
    case 'r':
    case 's':
    case 'd':
    case 'n'
  }
}
```

File: visualize/grpc/pb/visualize_grpc.pb.go [hunk 3]
```diff
@@ -1,96 +1,159 @@
+// Copyright (c) 2020, The OTNS Authors.
```
Comment (simonlingoogle) [inline visualize/grpc/pb/visualize_grpc.pb.go:1]: no copyright is needed for this file

File: visualize/grpc/pb/visualize_grpc.pb.go [hunk 4]
```diff
@@ -1,96 +1,159 @@
...
 
 type OtDeviceRole int32
 
 const (
-	OtDeviceRole_OT_DEVICE_ROLE_DISABLED OtDeviceRole = 0
-	OtDeviceRole_OT_DEVICE_ROLE_DETACHED OtDeviceRole = 1
-	OtDeviceRole_OT_DEVICE_ROLE_CHILD    OtDeviceRole = 2
-	OtDeviceRole_OT_DEVICE_ROLE_ROUTER   OtDeviceRole = 3
-	OtDeviceRole_OT_DEVICE_ROLE_LEADER   OtDeviceRole = 4
+	OtDeviceRole_OT_DEVICE_ROLE_DISABLED OtDeviceRole = 0 ///< The Thread stack is disabled.
```
Comment (simonlingoogle) [inline visualize/grpc/pb/visualize_grpc.pb.go:61]: this is automatically generated?

File: web/site/js/vis/Node.js [hunk 5]
```diff
@@ -140,12 +139,35 @@ export default class Node extends VObject {
...
+
+    _createPartitionSprite() {
+        return new PIXI.Sprite(this._getPartitionSpriteTexture());
+    }
+
+    _getStatusSpriteTexture() {
+        switch (this.role) {
+            case OtDeviceRole.OT_DEVICE_ROLE_LEADER:
+            case OtDeviceRole.OT_DEVICE_ROLE_ROUTER:
+                return Resources().WhiteSolidHexagon64.texture;
```
Comment (simonlingoogle) [inline web/site/js/vis/Node.js:153]: I feel that the Hexagon looks like it's a little smaller than Circle. Do you agree?
Should we increase the scale for Leader/Router for a little bit? @ctan-g @qilinkang

File: visualize/grpc/pb/visualize_grpc.pb.go (hunk 4 above)
Comment (ctan-g) [reply visualize/grpc/pb/visualize_grpc.pb.go:61]: Yeah, the whole `visualize_grpc.pb.go` got regenerated to proto version 4, including the copyright header.

File: web/site/js/vis/Node.js (hunk 5 above)
Comment (ctan-g) [reply web/site/js/vis/Node.js:153]: Maybe increase the radius to `22`?

File: types/types.go [hunk 6]
```diff
@@ -57,6 +57,31 @@ func DefaultNodeMode() NodeMode {
...
+			FullNetworkData = true
+		}
+	}
+	return NodeMode{
+		RxOnWhenIdle:       RxOnWhenIdle,
+		SecureDataRequests: SecureDataRequests,
+		FullThreadDevice:   FullThreadDevice,
+		FullNetworkData:    FullNetworkData,
+	}
+}
```
Comment (simonlingoogle) [inline types/types.go:83]:
```suggestion
func ParseNodeMode(s string) (mode NodeMode) {
	for _, c := range s {
		switch c {
		case 'r':
			mode.RxOnWhenIdle = true
		case 's':
			mode.SecureDataRequests = true
		case 'd':
			mode.FullThreadDevice = true
		case 'n':
			mode.FullNetworkData = true
		}
	}
	
	return 
}
```

Comment (ctan-g) [reply types/types.go:83]: Wow so magical

File: web/site/js/vis/Node.js (hunk 5 above)
Comment (simonlingoogle) [reply web/site/js/vis/Node.js:153]: Can you try and find a better radius? @ctan-g

Comment (ctan-g) [reply web/site/js/vis/Node.js:157]: `22` seems to make them look consistent.

Comment (simonlingoogle) [reply web/site/js/vis/Node.js:157]: Can you update the screenshot ?

Comment (ctan-g) [reply web/site/js/vis/Node.js:157]: Updated

Comment (simonlingoogle) [reply web/site/js/vis/Node.js:157]: Thanks, it looks better now.

File: web/site/js/vis/Node.js [hunk 7]
```diff
@@ -162,14 +188,37 @@ export default class Node extends VObject {
         this.label.text = this.id.toString() + "|" + rloc16
     }
 
+    _updateSize() {
+        var radius = CIRCULAR_SHAPE_RADIUS
```
Comment (simonlingoogle) [inline web/site/js/vis/Node.js:192]: Use `let` instead of `var`.

**Pull request merged**